src/main/scala/is/hail/sparkextras/ContextRDD.scala

```diff
@@ -0,0 +1,159 @@
+package is.hail.sparkextras
+
+import is.hail.io._
+import is.hail.expr.types._
+import is.hail.utils._
+
+import org.apache.spark._
+import org.apache.spark.rdd._
+import org.apache.spark.storage._
+import scala.reflect.ClassTag
+
+object ContextRDD {
+  def empty[C <: AutoCloseable, T: ClassTag](
+    sc: SparkContext,
+    mkc: () => C
+  ): ContextRDD[C, T] =
+    new ContextRDD(sc.emptyRDD[C => Iterator[T]], mkc)
+
+  // this one weird trick permits the caller to specify T without C
+  sealed trait Empty[T] {
+    def apply[C <: AutoCloseable](
+      sc: SparkContext,
+      mkc: () => C
+    )(implicit tct: ClassTag[T]
+    ): ContextRDD[C, T] = empty[C, T](sc, mkc)
+  }
+  private[this] object emptyInstance extends Empty[Nothing]
+  def empty[T] = emptyInstance.asInstanceOf[Empty[T]]
+
+  def union[C <: AutoCloseable, T: ClassTag](
+    sc: SparkContext,
+    xs: Seq[ContextRDD[C, T]]
+  ): ContextRDD[C, T] = {
+    assert(xs.length > 1)
+    val mkc = xs(0).mkc
+    assert(xs.tail.forall(_ == mkc))
+    union(sc, xs, mkc)
+  }
+
+  def union[C <: AutoCloseable, T: ClassTag](
+    sc: SparkContext,
+    xs: Seq[ContextRDD[C, T]],
+    mkc: () => C
+  ): ContextRDD[C, T] =
+    new ContextRDD(sc.union(xs.map(_.rdd)), mkc)
+
+  def weaken[C <: AutoCloseable, T: ClassTag](
```

Comment:
Why is this call weaken?  This just looks like a constructor that takes RDD[T] instead of RDD[C => Iterator[T]].

Reply:
In logic (and type theory), weakening is the structural rule that allows you to add unused hypothesis (or term variables) to the context of a proof (or term). Likewise, this adds an unused context to the `RDD[T]`.

Reply:
@patrick-schultz is on the nose. I forgot I had picked a wonky name. I'm open to other names. I'm fine with `apply`, though I think we have a tendency to have lots of constructors called `apply` which has lead me to default argument hell (b/c overloaded methods can't have defaults). As it stands that's not an issue here.